### PR TITLE
Do not recompile Kotlin buildscripts for non-ABI changes in classpath

### PR DESCRIPTION
**Pull request by rieske**
https://github.com/gradle/gradle/issues/9224

### Context
When buildscripts are compiled, they are cached and reused. The key components of the buildscript's cache key are the hash of content of the script itself and the hash of its classpath components. Thus, when either the buildscript or any of its classpath components change, we have a cache miss, the buildscript gets recompiled and cached with a new key.

The hash of jars on the buildscript's classpath depends on the content of the class files in those jars.
When buildscripts are compiled, they depend on the ABI (application binary interface) of the classpath which is a subset of the full classpath content. For example, the buildscript does not see private members nor the implementation details (method bodies) of visible members of the classpath - at compile-time it depends on the ABI. The bytecode of compiled buildscript remains valid if the classpath changes in a non-ABI breaking way.

To avoid recompiling buildscripts on non-ABI changes, the classpath hash that's part of the cache key can be computed against the ABI of the classpath instead of the full runtime classpath. This way, only ABI-breaking changes will cause a cache miss for the buildscript and for non-ABI changes, the compiled buildscripts would be reused from cache.

### Implementation

At high level, the `ClasspathHasher` is replaced with `CompileClasspathHasher` for computing the classpath hash component of the buildscript cache keys.

For the most part, [AbiExtractingClasspathResourceHasher](https://github.com/gradle/gradle/blob/3e1d8e2cebe108e657d21c65d6de5cbe511364b6/subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java) is reused to extract the Java ABI from classpath and compute its hash.

Here, however, we are compiling Kotlin buildscripts and Kotlin ABI is not the same as Java ABI.
Compiled Kotlin code adds Kotlin metadata to the class files that the Kotlin compiler reads and reacts to when compiling buildscripts. For example, Kotlin allows defining [inline functions](https://kotlinlang.org/docs/reference/inline-functions.html) that instruct Kotlin compiler to inline the method body to the call site. Thus, bodies of Kotlin inline functions have to be part of the ABI. 
The [internal](https://kotlinlang.org/docs/reference/visibility-modifiers.html) visibility modifier that's also encoded in Kotlin's metadata on the other hand, makes the class or method local to the module and invisible to the buildscripts and allows removing them from the ABI.

To cover the above cases, [KotlinAbiClassExtractor is injected into AbiExtractingClasspathResourceHasher](https://github.com/gradle/gradle/blob/2ba06e7134732525e79310b50e643e8e7e3f7944/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt#L134). It looks for Kotlin metadata annotations in class files, reads them and overrides the ABI generation from Java-based implementation to handle Kotlin-specific cases when needed.

#### Notes

- This is not "complete" compile avoidance for buildscripts - buildscripts will be recompiled for any ABI change on their classpath regardless if the given buildscript depends on the particular interface or not.

- In case we discover problems with this change after it is merged and want to exclude it from release, I added a [compile-time feature flag](https://github.com/gradle/gradle/blob/2ba06e7134732525e79310b50e643e8e7e3f7944/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/BuildServices.kt#L48) that can be flipped to disable compile avoidance for Kotlin buildscripts instead of reverting the whole thing.

### CODE REVIEW

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            }
+            return this
+        }
+
+        fun buildScriptNotCompiled(): OutputFixture {
+            if (output.contains(scriptClasspathCompileOperationStartMarker) || output.contains(scriptBodyCompileOperationStartMarker)) {
+                throw AssertionError("Expected script compilation to be avoided, but the buildscript was recompiled")
+            }
+            return this
+        }
```
Comment (rieske) [inline subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:230]: There has to be a better way to verify whether the build scripts were compiled or reused from cache than relying on `--debug` output.

**Comment by bamboo** [comment]
Great to see this happening 👍 

It should include tests for changes to inline functions as well.

**Comment by eskatos** [comment]
:rocket: 

It would also be good to have a test case where a non-abi change to a plugin that modifies the set of plugins it applies leads to the recompilation of the body of the scripts and available accessors in the body of the script change accordingly. For example, going from a precompiled script that applies a plugin to not applying it (without changing the abi) and asserting the accessors aren't available anymore.

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+        givenKotlinClassInBuildSrcContains(
+            """
+            inline fun t5() {
+                println("bar");
+            }
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("bar")
+    }
```
Comment (rieske) [inline subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:214]: Looks like Kotlin's `inline` functions is where this approach falls short. This test fails - the build script does not get recompiled and it uses the inlined code from first compilation.

Comment (bamboo) [reply subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:214]: Yeah, for that to work the classpath fingerprinter would need to be made aware of the Kotlin metadata for inline functions and include something like the bytecode of inline functions in the fingerprint.

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+import org.hamcrest.CoreMatchers
+import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+
+
+class BuildScriptCompileAvoidanceIntegrationTest : AbstractKotlinIntegrationTest() {
+
+    object CacheBuster {
```
Comment (bamboo) [inline subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:15]: ❓ Maybe the test should run with a custom script cache directory instead?

How about making `ScriptCacheRepository` accept a custom cache root directory via a system property the way `DefaultGeneratedGradleJarCache` does it? (see https://github.com/gradle/gradle/blob/5808250a35435bb95ec0b357374f50de1e65b1e8/subprojects/core/src/main/java/org/gradle/cache/internal/DefaultGeneratedGradleJarCache.java#L35)

The custom root directory could then be passed as a java.io.File scope here, https://github.com/gradle/gradle/blob/ae794708012c14b2dd636573d96bc5ae030e5d9c/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/ScriptCacheRepository.kt#L29

**Review by eskatos — COMMENTED** [review]
I wonder about `const val`s. They will be inlined as static finals. The base jvm abi extractor might already take care of that. A test with a `const val` changing value would be good.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt [hunk 4]
```diff
@@ -237,13 +237,14 @@ class StandardKotlinScriptEvaluator(
             scriptHost: KotlinScriptHost<*>,
             templateId: String,
             sourceHash: HashCode,
+            compilationClassPathHash: HashCode,
             parentClassLoader: ClassLoader,
```
[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt]: It is no longer used. Removed it.

[the remark this answers is not in the document]
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt]: Actually, I'll probably revert this one back and add a feature flag for now where we could switch this feature off without reverting the whole thing in case we find trouble with it before 6.8

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Assert.assertTrue
+import org.junit.Assume.assumeTrue
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.util.regex.Pattern
+
+
+class BuildScriptCompileAvoidanceIntegrationTest : AbstractKotlinIntegrationTest() {
```
Comment (big-guy) [inline subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:19]: Something I thought about today...

- I have a plugin that registers an extension with name `foo`. This plugin could be written in whatever language or come from a precompiled script plugin.
- I have a build script that applies the plugin and configures the extension by name. (e.g., `foo { /* configure */ }`)
- I run the build once and then change the plugin to register the extension with the name `bar`.
- I rerun the build, is the buildscript rebuilt?

I *think* this should cause us to generate new accessors and that means there's an ABI change. Is that what happens?

Comment (rieske) [reply subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt:19]: Confirmed that this is what happens with a test.

**Comment by rieske** [comment]
Rendered release note:
image

File: subprojects/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
 
 While the impact on your build may vary, most builds can expect a noticeably shorter feedback loop when editing Kotlin DSL build logic thanks to this improvement.
 
+### Compile avoidance for Kotlin DSL build scripts
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:52]: 💄 
```suggestion
### Compilation avoidance for Kotlin DSL scripts
```

File: subprojects/docs/src/docs/release/notes.md [hunk 7]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
...
+
+Until now, any changes to build logic in [buildSrc](userguide/organizing_gradle_projects.html#sec:build_sources) required all the build scripts in the project to be recompiled.
+
+This release introduces compile avoidance for [Gradle Kotlin DSL](userguide/kotlin_dsl.html) build scripts.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:56]: 💄 
```suggestion
This release introduces compilation avoidance for [Gradle Kotlin DSL](userguide/kotlin_dsl.html) scripts. 
```

File: subprojects/docs/src/docs/release/notes.md [hunk 8]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
 
 While the impact on your build may vary, most builds can expect a noticeably shorter feedback loop when editing Kotlin DSL build logic thanks to this improvement.
 
+### Compile avoidance for Kotlin DSL build scripts
+
+Until now, any changes to build logic in [buildSrc](userguide/organizing_gradle_projects.html#sec:build_sources) required all the build scripts in the project to be recompiled.
+
+This release introduces compile avoidance for [Gradle Kotlin DSL](userguide/kotlin_dsl.html) build scripts. 
+This feature will only cause the build scripts to be recompiled if a change to shared build logic changes the
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:57]: 💄 
```suggestion
This feature will cause the scripts to be recompiled only if a change to shared build logic impacts the 
```

File: subprojects/docs/src/docs/release/notes.md [hunk 9]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
...
+
+Until now, any changes to build logic in [buildSrc](userguide/organizing_gradle_projects.html#sec:build_sources) required all the build scripts in the project to be recompiled.
+
+This release introduces compile avoidance for [Gradle Kotlin DSL](userguide/kotlin_dsl.html) build scripts. 
+This feature will only cause the build scripts to be recompiled if a change to shared build logic changes the 
+ABI (application binary interface) of the resulting artifact.
+In simpler terms, changes to private implementation details of build logic, such as private methods or classes, 
+bodies of non-private methods or classes,
+as well as internal changes to [precompiled script plugins](userguide/custom_plugins.html#sec:precompiled_plugins) 
+such as configuration of tasks will no longer trigger recompilation of the project's build scripts.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:62]: ⭕ And that is awesome! :)

File: subprojects/docs/src/docs/release/notes.md [hunk 10]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
...
+
+While changes to `buildSrc` immediately affect the classpath of all the buildscripts, this improvement is more general and applies to changes in
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:69]: 💄 
```suggestion
While changes to `buildSrc` immediately affect the classpath of all the scripts, this improvement is more general and applies to changes in
```

File: subprojects/docs/src/docs/release/notes.md [hunk 11]
```diff
@@ -49,6 +49,27 @@ Garbage collection time goes from [2.6 seconds](https://scans.gradle.com/s/3bg67
...
+any jar on the buildscript's classpath that can be added by a plugin applied from an included build or added directly via `buildscript {}` block.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:70]: 💄 
```suggestion
any jar on the scripts classpath that can be added by a plugin applied from an included build or added directly via `buildscript {}` block.
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+    override fun writeClass(classMember: ClassMember, methods: Set<MethodMember>, fields: Set<FieldMember>, innerClasses: Set<InnerClassMember>) {
+        classMember.annotations.firstOrNull {
+            it.name == kotlinMetadataAnnotationSignature
+        }?.let {
+            when (val kotlinMetadata = KotlinClassMetadata.read(parseKotlinClassHeader(it))) {
+                is KotlinClassMetadata.Class -> kotlinMetadata.toKmClass().extractFunctionMetadata()
+                is KotlinClassMetadata.FileFacade -> kotlinMetadata.toKmPackage().extractFunctionMetadata()
+                else -> {
+                    // KotlinClassMetadata.SyntheticClass || KotlinClassMetadata.MultiFileClassFacade || KotlinClassMetadata.MultiFileClassPart || KotlinClassMetadata.Unknown
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:68]: ❓ I wonder if we should try to support more, fallback to compiling, or fail on some unsupported metadata types.

`MultiFileClassFacade` and `MultiFileClassPart` could happen if `@JvmMultifileClass` is used in build logic code.

`SyntheticClass` could happen if `@JvmSynthetic` is used in build logic code.

`Unknown` could happen if, for example, dependencies were compiled with a _future_ Kotlin version that introduce new metadata.

This could be addressed in a subsequent PR.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    override fun writeMethod(method: MethodMember) {
+        if (method.isInternal()) {
+            return
+        }
+
+        if (method.isInline()) {
+            inlineMethodWriter.writeMethod(method)
+        } else {
+            super.writeMethod(method)
+        }
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:89]: 💄 
```suggestion
        when {
            method.isInternal() -> return
            method.isInline() -> inlineMethodWriter.writeMethod(method)
            else -> super.writeMethod(method)
        }
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    private
+    fun KmDeclarationContainer.extractFunctionMetadata() {
+        this.extractInternalFunctions()
+        this.extractInlineFunctions()
+    }
+
+    private
+    fun KmDeclarationContainer.extractInlineFunctions() {
+        inlineFunctions.addAll(
+            this.functions
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:101]: ⭕ Could use sequence to create less intermediary lists

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+                .filter { Flag.Function.IS_INLINE(it.flags) }
+                .mapNotNull { it.signature?.asString() }
+        )
+    }
+
+    private
+    fun KmDeclarationContainer.extractInternalFunctions() {
+        internalFunctions.addAll(
+            this.functions
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:110]: ⭕ Could use sequence to create less intermediary lists

**Review by eskatos — CHANGES_REQUESTED** [review]
Great work! This will make the Kotlin DSL more relevant and alleviate pain, can't wait to use it daily 👍 

Kudos for the good coverage.

Left some requested changes ❌, comments ⭕, questions ❓ and suggestions 💄.

**Comment by big-guy** [comment]
@rieske I'm going to merge this as-is and get a nightly built, please follow-up on @eskatos comments in a separate PR.

**Comment by rieske** [comment]
Separate PR for ironing out the release note: https://github.com/gradle/gradle/pull/14820

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt (hunk 13 above)
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:89]: Handling this one here https://github.com/gradle/gradle/pull/14821

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt (hunk 15 above)
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:110]: Handling this one here https://github.com/gradle/gradle/pull/14821

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt (hunk 14 above)
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:101]: Handling this one here https://github.com/gradle/gradle/pull/14821

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt (hunk 12 above)
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:68]: Here's a fix for `MultiFileClassFacade` and `MultiFileClassPart` - missed one case there - https://github.com/gradle/gradle/pull/14827

Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt:68]: And for `Unknown` metadata - fallback to using the full hash of the artifact: https://github.com/gradle/gradle/pull/14834
